master/internal/api_runs.go

```diff
@@ -503,3 +504,135 @@ func (a *apiServer) KillRuns(ctx context.Context, req *apiv1.KillRunsRequest,
 	}
 	return &apiv1.KillRunsResponse{Results: results}, nil
 }
+
+func (a *apiServer) DeleteRuns(ctx context.Context, req *apiv1.DeleteRunsRequest,
+) (*apiv1.DeleteRunsResponse, error) {
+	curUser, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+	// Get active runs and kill
+	var activeRuns []int32
+	activeStates := []model.State{model.ActiveState, model.PausedState, model.RunningState}
+
+	killQuery := db.Bun().NewSelect().
+		ModelTableExpr("runs AS r").
+		Model(&activeRuns).
+		Column("r.id").
+		Join("LEFT JOIN trials_v2 t ON r.id=t.run_id").
+		Join("LEFT JOIN experiments e ON r.experiment_id=e.id").
+		Join("JOIN projects p ON r.project_id = p.id").
+		Join("JOIN workspaces w ON p.workspace_id = w.id").
+		Where("r.state IN (?)", bun.In(activeStates)).
+		Where("r.project_id = ?", req.ProjectId)
+
+	if req.Filter == nil {
+		killQuery = killQuery.Where("r.id IN (?)", bun.In(req.RunIds))
+	} else {
+		killQuery, err = filterRunQuery(killQuery, req.Filter)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	err = killQuery.Scan(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	killReq := apiv1.KillRunsRequest{
+		RunIds:    activeRuns,
+		ProjectId: req.ProjectId,
+	}
+	killResp, err := a.KillRuns(ctx, &killReq)
```

Comment:
If delete runs is going to kill runs, then I think this API needs to be async/backgrounded. Kill is not an instant action, and killing a trial then deleting it from the DB instantly will cause issues ("undefined behavior".. I'm not sure what will happen but it's not on purpose and probably bad).

Why do we need to kill active runs rather than only allowing delete for terminal runs and returning "400 cannot delete active runs"? I understand it's a marginally better UX, but it is at the cost of a lot of complexity.

Reply:
There was a misunderstanding on my part, I believed that we should have been killing a run on deletion. Removing this logic.